### ISSUE TITLE
UFunc userloop selection does not work when only output is void/user dtype

### ISSUE THREAD

**Issue by mhvk** [numpy/numpy#11109]
I've been trying to wrap functions into ufuncs and ran into a problem for some that have regular doubles as inputs, but a structured dtype as output: if I follow the instructions and define a userloop for these [1], it works if I explicitly pass in an output with the right `dtype`, but not if I don't.

After digging into the code, I found the problem arises because `linear_search_userloop_type_resolver` only attemps to see whether a user loop should be looked at if one of the operands has a user-defined or void type [2]. Hence, if one does not give an output, the search will not done.

This does not seem very difficult to fix, but I'm not sure whether it would break expectations elsewhere.

[1] https://docs.scipy.org/doc/numpy/user/c-info.ufunc-tutorial.html#example-numpy-ufunc-with-structured-array-dtype-arguments
[2] https://github.com/numpy/numpy/blob/master/numpy/core/src/umath/ufunc_type_resolution.c#L1742

**Comment by mhvk** [comment]
@mattip - that seems wonderfully far-fetched! It sounds like it would make sense to make a PR... (I'll try to do that after finishing my ufunc-wrapping project; for now, I'm simply setting their `ufunc->type_resolver` to a routine that works around it.)

**Comment by seberg** [comment]
Not that it (hopefully) matters much anymore. But my expectation is that the largest issue to just adding something like that would be that if there are any usertypes it is probably a non-negligible overhead in the current setup.

**Comment by mhvk** [comment]
@seberg - the check for the presence of a user loop is very fast (just whether a pointer is NULL or not) - any checks would only have to be done if user loop is present.

**Comment by seberg** [comment]
This now works if the loops match exactly.  You need promoter if not, that also works although the API may still be a bit in flux.

**Comment by mhvk** [comment]
@seberg - last November, you commented above that this issue should now be solved, so I thought I would try to see if our ERFA routines could now do without a custom loop selector. Unfortunately, I still get the same failures as before, of being unable to find a loop if my input types have standard dtypes, but my output is a structured dtype (and I do not pass in an output). 

It still seems a bit odd that anything more would be needed here; surely, if no output is given, loop selection should ignore the output dtype since the output will be created with the correct dtype anyway.

Am I missing something? If this is not something that should "just work", how should this case be handled? In your comment, you mentioned a possible need for a promotor, but what exactly would that need to do? Is there an example I could look at?

**Comment by seberg** [comment]
Hmmm, I would expect loop selection to work, but if and only if the signature matches exactly (without the output) and there is no competing signature.  I.e. the DType class of inputs must be exact, which for integers might be tricky in practice (since there are multiple equivalent ones).

But I think there is another piece anyway?
1. We need to find the loop, this should just work (at the second call or in the exact conditions above, but maybe I am missing a path where this isn't ensured).
2. You need the output dtype instance.  That is always the same, I suspect, but it is an (arbitrary) instance of the Void dtype class.  That still needs to call a `resolve_dtypes`, and that still needs to call the legacy type resolver in this case.

So that same error *might* also be the second call (I don't know)?

Preliminary API?  Yes, a promoter function (if necessary) to tweak the input DTypes (classes/types) so that the loop matches exactly.  And the current `resolve_dtypes` lives on the loop and has a different signature.

**Comment by mhvk** [comment]
@seberg - I wondered if I was too quick to say things failed, because in many test cases I just input python float and int, but it does not seem to be the case:
```
In [1]: import numpy as np

In [2]: np.__version__
Out[2]: '1.25.0.dev0+941.gb35aac2c3'

In [3]: from erfa import ufunc as erfa_ufunc

In [4]: erfa_ufunc.epv00(np.array([2400000.5]), np.array([53411.52501161]))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[4], line 1
----> 1 erfa_ufunc.epv00(np.array([2400000.5]), np.array([53411.52501161]))

TypeError: ufunc 'epv00' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

erfa_ufunc.epv00.resolve_dtypes((np.dtype('f8'), np.dtype('f8'), None, None, None))
TypeError: ufunc 'epv00' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

But perhaps I am simply doing something wrong. Currently, the function is defined with
```
    ufunc = (PyUFuncObject *)PyUFunc_FromFuncAndDataAndSignature(
        NULL, NULL, NULL,
        0, 2, 3, PyUFunc_None,
        "epv00",
        "UFunc wrapper for eraEpv00",
        0, NULL);
    if (ufunc == NULL) {
        goto fail;
    }
    dtypes[0] = dt_double;
    dtypes[1] = dt_double;
    dtypes[2] = dt_pv;
    dtypes[3] = dt_pv;
    dtypes[4] = dt_int;
    status = PyUFunc_RegisterLoopForDescr(
        ufunc, dt_pv,
        ufunc_loop_epv00, dtypes, NULL);
    if(status != 0){
        Py_DECREF(ufunc);
        goto fail;
    }
```
where in this trial I removed setting our own resolver,
```
ufunc->type_resolver = &ErfaUFuncTypeResolver;
```
```

**Comment by seberg** [comment]
No, there is nothing you can do, without using the "experimental DType API".  Because I think it is correct, the first step "finding the loop" probably succeeds, but it works with "d,d->V" on the dtype types only.  Only the type resolver knows what the correct void/structured instance is.

Now, looking at this these "user loops", you actually do pass exact dtype *instances* in at creation time, so NumPy has that information in principle.

So one thing that you/we could do is tag those on somewhere to the `ArrayMethod` (or look it up) to force exactly those dtypes to be used in the inner-loop.

**Comment by mhvk** [comment]
Yes, I guess there are two options, one to make the numpy code able to deal with structured dtypes (which are not really user types in that for structured dtypes numpy knows how the data is stored), and the other to treat this as true user types (opague blobs for numpy).

In case you have the time, it might be useful to have a look at my replacement [type resolver](https://github.com/liberfa/pyerfa/blob/c12b70e84a9cf3bd4407d39412c111ad6f62972a/erfa/ufunc.c.templ#L532) and my replacement for [`ufunc_loop_matches`](https://github.com/liberfa/pyerfa/blob/c12b70e84a9cf3bd4407d39412c111ad6f62972a/erfa/ufunc.c.templ#L334).

Looking back at those myself, I see that another thing I had to "fix" is to make sure that strings and user dtypes with a single string are considered safe for casting purposes. Though the reason for even having the user dtype with a single string element was that that was the only way to make it a user type and thus allow userloops...

**Comment by seberg** [comment]
@mhvk I am a bit unsure where this is going.  The new API is:
1. You register this as a `d,d->V` ArrayMethod/loop (all classes, so its just structured).  This loop will be found (if it matches exactly).
2. The loop itself has a `resolve_dtypes` function that does (part) of the job of your type resolver, but more narrow/well defined.
3. If you want e.g. a `d,d->V` loop to also kick in for `i,i->V` or so, you may need a "promoter", a function similar to `promote((i, i, None)) -> (d, d, V)`.  (I am very much open to tweak this).

With the current API, you cannot avoid your replacement type resolver without adding a shim to make it work for you (which I think is possible), because you need it for step 2.

**Comment by mhvk** [comment]
@seberg - despite reviewing your PRs, I'm still not very up to snuff on how to actually use the new API, so thanks for that summary!

I think what would probably help me most at this stage is just to be pointed to documentation or an example of doing those three steps - I can then adjust (and perhaps add a bit of documentation specific for structured dtype).
